### ISSUE TITLE
conflicting types for 'tuntap_open' （Error in cross compiling Windows version with MinGW in Linux）

### ISSUE THREAD

**Issue by llfj** [ntop/n2n#631]
use  MinGW for cmake cross compiling in linux

1 error：
/n2n/include/random_numbers.h:47:10: fatal error: Wincrypt.h: No such file or directory
  47 | #include <Wincrypt.h>   // HCTYPTPROV, Crypt*-functions
 To  #include <wincrypt.h>

------------------------------------------------------------------------------------------
2 error：
/n2n/win32/wintap.h:10: warning: "_CRT_SECURE_NO_WARNINGS" redefined
   10 | #define _CRT_SECURE_NO_WARNINGS
      | 
In file included from /n2n/include/minilzo.h:52,
                 from /n2n/include/n2n.h:125,
                 from /n2n/win32/wintap.c:5:
/n2n/include/lzodefs.h:113: note: this is the location of the previous definition
  113 | #define _CRT_SECURE_NO_WARNINGS 1
      | 
**/n2n/win32/wintap.c:408:5: error: conflicting types for 'tuntap_open'
  408 | int tuntap_open(struct tuntap_dev *device,**
      |     ^~~~~~~~~~~
In file included from /n2n/win32/wintap.c:5:

I Will not modify。。。。。。。。。。。。。
-----------------------------------------------------------------------------------------------------

**Comment by Logan007** [comment]
Did MinGW CMake build ever work or is this a first time?

Anyone experienced in this?

**Comment by llfj** [comment]
Netizens say there are errors in n2n.h。
v2.7.0 V2.8.0 Can be compiled normally。

**Comment by Logan007** [comment]
So, only 2.9 fails?

2.7 and 2.8 work?

By the way, MSVC [building instructions](doc/Building.md) for Windows work very well; in case you want to give that a try.

**Comment by llfj** [comment]
I always use cmake cross compiler under linux.

V2.7.0 r525 is ok
-----------------------------------------------------------
Scanning dependencies of target example_sn_embed
[ 93%] Building C object CMakeFiles/example_sn_embed.dir/src/example_sn_embed.c.obj
Linking C executable example_sn_embed.exe
[ 93%] Built target example_sn_embed
Scanning dependencies of target n2n-benchmark
[ 96%] Building C object CMakeFiles/n2n-benchmark.dir/tools/benchmark.c.obj
/root/n2n2/tools/benchmark.c: In function 'run_transop_benchmark':
/root/n2n2/tools/benchmark.c:138:3: warning: implicit declaration of function 'gettimeofday'; did you mean 'mingw_gettimeofday'? [-Wimplicit-function-declaration]
  138 |   gettimeofday( &t1, NULL );
      |   ^~~~~~~~~~~~
      |   mingw_gettimeofday
Linking C executable n2n-benchmark.exe
[ 96%] Built target n2n-benchmark
Scanning dependencies of target supernode
[100%] Building C object CMakeFiles/supernode.dir/src/sn.c.obj
Linking C executable supernode.exe
[100%] Built target supernode
[root@centos7 cmake]# 
[root@centos7 cmake]# git rev-list --count HEAD
525
---------------------------------------------------------------

**Comment by Logan007** [comment]
Thank you for reporting. We might need help from some expert here.

Anyone?

**Comment by llfj** [comment]
I compiled successfully，I annotated the tuntap_open in n2n.h

/* int tuntap_open (tuntap_dev *device, char *dev, const char *address_mode, char *device_ip,
                 char *device_mask, const char * device_mac, int mtu);
int tuntap_read (struct tuntap_dev *tuntap, unsigned char *buf, int len);
int tuntap_write (struct tuntap_dev *tuntap, unsigned char *buf, int len);
void tuntap_close (struct tuntap_dev *tuntap);
void tuntap_get_address (struct tuntap_dev *tuntap); */
----------------------------------------------------------------------------------------------
I don’t understand C, can it be modified as follows：

#ifndef WIN32
int tuntap_open (tuntap_dev *device, char *dev, const char *address_mode, char *device_ip,
                 char *device_mask, const char * device_mac, int mtu);
int tuntap_read (struct tuntap_dev *tuntap, unsigned char *buf, int len);
int tuntap_write (struct tuntap_dev *tuntap, unsigned char *buf, int len);
void tuntap_close (struct tuntap_dev *tuntap);
void tuntap_get_address (struct tuntap_dev *tuntap);
#endif

**Comment by Logan007** [comment]
> I compiled successfully

That's great news! Do you want to provide short instructions how to compile using MinGw? We could add it to our [building documentation](https://github.com/ntop/n2n/blob/dev/doc/Building.md).

> #ifndef WIN32

If have just checked if it still gets compiled by MSVC. And yes, it does. So, I guess we can adopt this change to the code. Do you want to provide a pull request?

Everybody: Any other side effects observed?

**Comment by pvagner** [comment]
Hello,

I'm on arch linux, I've just cross-compiled current master head for Windows and tested it on a Windows 10 virtual machine. It's working fine for me. I am only using edge.exe so far.

This is my setup and steps:
* I've installed these packages from the AUR: mingw-w64-binutils-bin mingw-w64-crt-bin mingw-w64-environment mingw-w64-gcc-bin mingw-w64-headers-bin mingw-w64-openssl mingw-w64-winpthreads-bin mingw-w64-zlib . I am not sure zlib is needed but I have it installed as I was building something else in the past.
* Then I've just cloned this repo.
* Patched the code as per this issue
* Borrowed the windows specific cmake toolchain file from here: https://github.com/meyerd/n2n/blob/master/n2n_v2/cmake/CMakeToolchainFileMingw32.cmake
* Edited a few variables at the top:
	```# which compilers to use for C and C++
	SET(CMAKE_C_COMPILER x86_64-w64-mingw32-gcc)
	SET(CMAKE_CXX_COMPILER x86_64-w64-mingw32-g++)

	# here is the target environment located
	SET(CMAKE_FIND_ROOT_PATH /usr/x86_64-w64-mingw32/)```
* Created a folder windows-build and changed into it:
	```mkdir windows-build
	cd windows-build```
* Executed cmake like this:
	```cmake -DCMAKE_TOOLCHAIN_FILE=../cmake/CMakeToolchainFileMingw32.cmake --build ./ ../```
* For some more info see the wiki at: https://github.com/meyerd/n2n/wiki . Of course parts of it are outdated but these cross compile instructions are still applicable.

**Comment by Logan007** [comment]
Thank you!

I will try to compile all that (and even try to follow it once, too) and create an entry to our documentation. I hope it will be helpful to have an alternative to MSVC. As a reminder to myself, I will re-open this issue @llfj 

@pvagner, do you think that our CMake file requires permanent changes? I am not too familiar with CMake. Please feel free to open a pull request with your changes and also an additional section to out `doc/Building.md` documentation.

**Comment by llfj** [comment]
QQ group of our friends have been fixed it .

vtcpip fix windows compile (#636)
b42c111
18 hours ago

**Comment by Logan007** [comment]
Oh, this is wonderful! Greetings to the group! :orange_heart: 

I just have not realized that it was related.

I will close as soon as instructions to `doc/Building.md` have been added.
